docs/sources/clients/promtail/logrotation.md

```diff
@@ -0,0 +1,137 @@
+
+One of the critical components here is the log rotater. Let's understand how it impacts other components like the appender and tailer.
+
+In general, when a log is rotated, it can be done in two ways.
+
+Given a log file `error.log`
+
+1. Copy the log file with a different name e.g `error.log.1` and truncate the original log file `error.log`.
+2. Rename the log file with a different name e.g: `error.log.1` and create a new log file with the original name `error.log`.
+
+In both cases, after log rotation, all the new log lines go to the original `error.log` file.
+
+These types are shown in the images below.
+
+### Copy and Truncate
+![block_diagram](./logrotation1.png)
+
+### Rename and Create
+![block_diagram](./logrotation2.png)
+
+Both types of log rotation seem to give the same result. However, there are some subtle differences.
+
+(1) favors the `appender` as its descriptor for the original log file `error.log` doesn't change. Therefore, it can keep writing to the same file descriptor. In other words, re-opening the file is not needed.
+
+However (1) has a serious problem if we bring the `tailer` into account. There is a race between truncating the file and the `tailer` finishing reading that log file. Meaning, there is a high chance of the log rotation mechanism truncating the file `error.log` before the `tailer` reads everything from it.
+
+This is where (2) can help. Here, when the log file `error.log` is renamed to `error.log.1`, the `tailer` still holds the file descriptor of `error.log.1`. Therefore, it can continue reading the log file until it's completed. But that comes with the tradeoff: with (2), you have to signal the `appender` to reopen `error.log` (and `appender` should be able to reopen it). Otherwise, it would keep writing to `error.log.1` as the file descriptor won't change. The good news is that most of the popular `appender` solutions (e.g: Syslog, Kubelet, Docker log driver) support reopening log files when they are renamed.
+
+We recommend (2) as that is the one which works well with Promtail (or any similar log scraping agent) without any data loss. Now let's understand how we exactly configure log rotation in different platforms.
+
+## Configure log rotation.
+
+Your logs can be rotated by different components depending on where you are running your application or services. If you are running on Linux bare metal or Linux VMs, high chance you will be using [`logrotate`](https://man7.org/linux/man-pages/man8/logrotate.8.html) utility. However, if you are running in Kubernetes, it's not that obvious who rotates the logs and interestingly it may depend on what container runtime your Kubernetes cluster is using.
```

Comment:
```suggestion
Your logs can be rotated by different components depending on where you are running your applications or services. If you are running them on Linux bare metal or Linux VMs, there is a high chance you will be using the [`logrotate`](https://man7.org/linux/man-pages/man8/logrotate.8.html) utility. However, if you are running them in Kubernetes, it's not that obvious who rotates the logs and interestingly it may depend on what container runtime your Kubernetes cluster is using.
```

Reply:
Is there a difference between "Linux bare metal" and "Linux VMs"?